Recreated types for flux@3.2

GiedriusGrabauskas opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

andy-ms commented on types/flux/utils.d.ts:
```diff
@@ -0,0 +1,17 @@
+// Type definitions for Flux Utils 3.1.2
```
I thought the linter should have caught this, but only `index.d.ts` should have a header.

GiedriusGrabauskas commented:
Updated.

RyanCavanaugh commented:
Something is making the Travis CI not run on this PR - maybe you can force push a squash commit to get it going again?

GiedriusGrabauskas commented:
Squashed and fixed CI tests.

GiedriusGrabauskas commented:
Waiting for merge. @andy-ms

andy-ms commented:
Will this be backwards-compatible with the old types? What changes were made besides breaking things into separate files?

GiedriusGrabauskas commented:
Okey, I fixed some incorrect implementation. But now it's fully backwards-compatible with the old types.

The main change was made for `FluxContainer` method `create`.
Before:
```ts
function create<TComponent>(base: React.ComponentClass<TComponent>, options?: RealOptions): React.ComponentClass<TComponent>;
```
Now:
```ts
function create<TProps>(base: Component<TProps, any, any>, options?: RealOptions): Component<TProps, any, any>;
function create<TProps, TState>(base: Component<TProps, TState, any>, options?: RealOptions): Component<TProps, TState, any>;
function create<TProps, TState, TContext>(base: Component<TProps, TState, TContext>, options?: RealOptions): Component<TProps, TState, TContext>;
function create<TProps, TState, TContext, TStatic>(base: Component<TProps, TState, TContext> & TStatic, options?: RealOptions): Component<TProps, TState, TContext> & TStatic;
```

andy-ms commented:
So, what did this pull request change? Why was it necessary to re-create the types?

GiedriusGrabauskas commented:
# Full changelog
- Added missing properties for `FluxStore` class ([source](https://github.com/facebook/flux/blob/master/src/stores/FluxStore.js#L31)):
   - [__emitter](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-609ecf5564f8baae6365b3fbcf08f2acR71)
   - [__dispatcher](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-609ecf5564f8baae6365b3fbcf08f2acR69)
   - [__className](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-609ecf5564f8baae6365b3fbcf08f2acR67)
   - [_changeEvent](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-609ecf5564f8baae6365b3fbcf08f2acR65)
   - [__changed](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-609ecf5564f8baae6365b3fbcf08f2acR63)
   
- Added protection for class `FluxStore` methods and properties:
   - [__onDispatch](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-609ecf5564f8baae6365b3fbcf08f2acR61)
   - [__invokeOnDispatch](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-609ecf5564f8baae6365b3fbcf08f2acR54)
   - [__emitChange](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-609ecf5564f8baae6365b3fbcf08f2acR47)

- Added static methods for `FluxContainer` component ([ComponentStatic](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-8ecd2f07c2acb848c292374a564111b4R40)).

- Added `tslint.json` and fixed all errors.

- Created folders structure from [source repo](https://github.com/facebook/flux/tree/master/src).

- Added more generics for `create` method in `FluxContainer` ([changes](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-8ecd2f07c2acb848c292374a564111b4R52)).

- [Created](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-e575db0fb790831e830e7b5f8b75ff96) and [exported](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-ea74635536f037349ee57a5e785f1f15R8) `Mixin` function from `flux/utils` ([source](https://github.com/facebook/flux/blob/master/utils.js#L11)).

- Added missing `abstract` in `ReduceStore` methods ([source](https://github.com/facebook/flux/blob/master/src/stores/FluxReduceStore.js#L63)):
   - [getInitialState](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-25847b7c9d3f68a60fa943524c0b07bdR20)
   - [reduce](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-25847b7c9d3f68a60fa943524c0b07bdR27)

- Changed `FluxReduceStore` to [abstract class](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-25847b7c9d3f68a60fa943524c0b07bdR9).

- `Function` changed to `()=> void` ([#diff-609ecf5564f8baae6365b3fbcf08f2acR22](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15800/files#diff-609ecf5564f8baae6365b3fbcf08f2acR22)).
